### PR TITLE
Add end-to-end HTTP usecase

### LINKED ISSUE #33: First end-to-end sample (HTTPClient -> NIO Server -> ...)
First sample in which we want to:

- have a client issue an http request (our "wrapped" http client for now)
  - a) with some values in context
  - b) without any values in context
- `inject` from context to headers
- receive on a server and `extract` the context
  - show that we got the same values 
  - show how an inject could "make up" a new trace id if none was present in incoming request
   - fun topic: showcase how we'd implement sampling here ("one in n requests should be traced") etc
- the server should log something, and these logs should include the metadata we got from the context

Maybe make another request from there?

### PR DESCRIPTION
**Pull request by slashmo**
Closes #33

### CODE REVIEW

File: UseCases/Sources/HTTPEndToEnd/main.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import NIO
+import NIOHTTP1
+
+import NIO
+
+let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: System.coreCount)
+let threadPool = NIOThreadPool(numberOfThreads: 6)
+threadPool.start()
+
+let orderServiceBootstrap = ServerBootstrap(group: eventLoopGroup)
```
Comment (ktoso) [inline UseCases/Sources/HTTPEndToEnd/main.swift]: Minor:

would be nice to keep the noise out of the "first" lines, so we could:

```
let orderServiceBootstrap = serviceBootstrap(handler:
    OrderServiceHandler(httpClient: httpClient, instrument: FakeTracer())
)
let storageServiceBootstrap = serviceBootstrap(handler:
    StorageServiceHandler()
)
```

with the setup lower in the file (at the end):

```

// MARK: - Setups

func serviceBootstrap(handler: ChannelHandler) -> ServerBootstrap {
    ServerBootstrap(group: eventLoopGroup)
        .serverChannelOption(ChannelOptions.backlog, value: 256)
        .serverChannelOption(ChannelOptions.socketOption(.so_reuseaddr), value: 1)
        .childChannelInitializer { channel in
            channel.pipeline.configureHTTPServerPipeline().flatMap {
                channel.pipeline.addHandler(handler)
            }
        }
        .childChannelOption(ChannelOptions.socketOption(.so_reuseaddr), value: 1)
        .childChannelOption(ChannelOptions.maxMessagesPerRead, value: 1)
        .childChannelOption(ChannelOptions.allowRemoteHalfClosure, value: true)
}


```

Comment (slashmo) [reply UseCases/Sources/HTTPEndToEnd/main.swift]: Good idea 👌

File: UseCases/Sources/HTTPEndToEnd/Services/OrderServiceHandler.swift [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import Logging
+import NIO
+import NIOHTTP1
+
+final class OrderServiceHandler: ChannelInboundHandler {
+    typealias InboundIn = HTTPServerRequestPart
+    typealias OutboundOut = HTTPServerResponsePart
+
+    private let httpClient: InstrumentedHTTPClient
+    private let instrument: Instrument<HTTPHeaders, HTTPHeaders>
```
Comment (ktoso) [inline UseCases/Sources/HTTPEndToEnd/Services/OrderServiceHandler.swift:14]: Yup good :) 

Soon let's also tackle: https://github.com/slashmo/gsoc-swift-tracing/issues/47

File: UseCases/Sources/HTTPEndToEnd/Services/OrderServiceHandler.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        I.InjectInto == HTTPHeaders,
+        I.ExtractFrom == HTTPHeaders {
+        self.httpClient = httpClient
+        self.instrument = Instrument(instrument)
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        guard case .head(let requestHead) = self.unwrapInboundIn(data) else { return }
+
+        var baggage = BaggageContext()
```
Comment (ktoso) [inline UseCases/Sources/HTTPEndToEnd/Services/OrderServiceHandler.swift:28]: thats actually an interesting case of why calling it context would have been annoying... here we'd very confusingly nameclash hm hm.

Looks good!

File: UseCases/Sources/HTTPEndToEnd/Services/StorageServiceHandler.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        guard case .head(let requestHead) = self.unwrapInboundIn(data) else { return }
+
+        var baggage = BaggageContext()
+        baggage[BaggageContext.BaseLoggerKey.self] = self.logger
+        self.instrument.extract(from: requestHead.headers, into: &baggage)
+
+        baggage.logger.info("📦 Looking for the product")
+
+        sleep(2)
```
Comment (ktoso) [inline UseCases/Sources/HTTPEndToEnd/Services/StorageServiceHandler.swift]: 🛑 😉 

How about this:

```suggestion
        context.eventLoop.scheduleTask(in: .seconds(2)) {
            baggage.logger.info("📦 Found the product")
            
            let responseHead = HTTPResponseHead(version: requestHead.version, status: .ok)
            ...
```

**Review by ktoso — APPROVED** [review]
LGTM, let's avoid the sleep in the handler :)

**Pull request merged**